### DISCUSSION TITLE
Make an async call within main.ts to set routes?

### DISCUSSION THREAD

**Discussion by Jumborner** [frandiox/vite-ssr#92] [Q&A] [answered]
Hello, I've been trying to implement vite-ssr into my app. It works when I set the routes import manually, but what I need is to set the routes depending on the current domain. What I have:

`const resourceId =  await getDomainId(domain);
    if (resourceId == "000000000000000000" || resourceId == "11111111111111111") {
  routes =  routes1;
} else if (resourceId == "333333") {
  routes =   routes2
} else {
  routes = routes2
}`

I have been unable to call the getDomainId() however. I have tried making an async function and then running and moving it all to a service and importing it. I keep getting an 'Can not use keyword 'await' outside an async function' error. Funny thing is, with the normal router approach in vite my solution had previously worked ( when I made the router, imported it and had App.use(router).

With vite ssr I can't figure out how to do this =/

**Comment by frandiox** [comment, marked as answer]
Can you write your whole `main.js`/entry code?
You should be able to add routes to the router in the main hook like this: https://github.com/frandiox/vite-ssr/discussions/32#discussioncomment-823070

**Reply by Jumborner** [reply]
> Can you write your whole `main.js`/entry code?
> You should be able to add routes to the router in the main hook like this: [#32 (comment)](https://github.com/frandiox/vite-ssr/discussions/32#discussioncomment-823070)

Hey. I've managed to make the call using `(async () => {
   domain = getDomain()
  try {
      resourceId = await getDomainId(domain);
      
  } catch (e) {
      console.log(e)
  }
})();
My new issue is  that, the getDomain method tries to get the current domain from the url. 
```
export function getDomain(): any {
  
  if (typeof window !== "undefined") {
    return window.location.host;
  } 
}
```

So now Im thinking on how to get it for the SSR. Like I've said, I need to change the routes depending on the domain.

**Reply by frandiox** [reply]
The `url` is in the SSRContext and passed to vite-ssr main hook.

**Reply by Jumborner** [reply]
> The `url` is in the SSRContext and passed to vite-ssr main hook.

Yeah, but I dont think that is what I need. What I need is to determine which "routes" array to pass to `export default viteSSR(App, { **routes** }, async ({ app, router, isClient, url, initialState, initialRoute, request }) => ` 

I have several possible domains that the user can use. Depending on which one it is, I want to change the routes. Before SSR i did a check with `window.location.host;` checking which domain it was, and then decided which routes to use.

**Reply by frandiox** [reply]
Yeah I think I understand what you want to make but we don't have any async hook before the actual `vite-ssr` main hook.
What I'm saying is that other users have added routes in this main hook successfully so perhaps you can try it.
The router doesn't check the URL until the main hook is resolved, so you can add routes there.

From https://github.com/frandiox/vite-ssr/discussions/32#discussioncomment-823070 :

```js
import viteSSR from 'vite-ssr'
import { addPagePropsGetterToRoutes } from 'vite-ssr/vue/utils'

export default viteSSR(
  App,
  {
    routes: [] // leave this empty or put here the shared routes
  },
  async({ url, router }) => {
    // main hook

    let domainDependentRoutes = []
    if (domainCondition(url)) {
      domainDependentRoutes = [a,b,c]
      // you can also import them asyncrhonously: await import('./routes-for-X-domain')
    } else {
      domainDependentRoutes = [d,e,f]
    }

    // This adds vite-ssr stuff to the routes manually
    addPagePropsGetterToRoutes(domainDependentRoutes)
    // This adds routes to Vue Router
    domainDependentRoutes.forEach(router.addRoute)

  })
```

Tell me if this works for you.

**Reply by Jumborner** [reply]
> Yeah I think I understand what you want to make but we don't have any async hook before the actual `vite-ssr` main hook.
> What I'm saying is that other users have added routes in this main hook successfully so perhaps you can try it.
> The router doesn't check the URL until the main hook is resolved, so you can add routes there.
> 
> From [#32 (comment)](https://github.com/frandiox/vite-ssr/discussions/32#discussioncomment-823070) :
> 
> ```js
> import viteSSR from 'vite-ssr'
> import { addPagePropsGetterToRoutes } from 'vite-ssr/vue/utils'
> 
> export default viteSSR(
>   App,
>   {
>     routes: [] // leave this empty or put here the shared routes
>   },
>   async({ url, router }) => {
>     // main hook
> 
>     let domainDependentRoutes = []
>     if (domainCondition(url)) {
>       domainDependentRoutes = [a,b,c]
>       // you can also import them asyncrhonously: await import('./routes-for-X-domain')
>     } else {
>       domainDependentRoutes = [d,e,f]
>     }
> 
>     // This adds vite-ssr stuff to the routes manually
>     addPagePropsGetterToRoutes(domainDependentRoutes)
>     // This adds routes to Vue Router
>     domainDependentRoutes.forEach(router.addRoute)
> 
>   })
> ```
> 
> Tell me if this works for you.

Hi, thanks for your reply. Your solution seems to be working :), as it allowed me to move forward ( its loading different things depending on the domain, although some issues with SSR and components for some reason, but this is my next to-do ).
